### PR TITLE
Documentation: FieldTimeSeries, FieldDataset and basic post-processing

### PR DESCRIPTION
**Pull request by erin-atkinson**
I've added some documentation for the **FieldTimeSeries & Post-Processing** tutorial entry. This includes `FieldTimeSeries`, `FieldDataset` and a example offline processing pipeline. Hopefully the formatting is all okay? I tried to copy what other manual sections look like.

### CODE REVIEW

File: docs/src/field_time_series.md [hunk 1]
```diff
@@ -1 +1,403 @@
...
+    └── max=0.0, min=0.0, mean=0.0
+```
+
+Additionally, a `FieldTimeSeries` may be generated from simulation output:
+
+```jldoctest field_time_series
+model = NonhydrostaticModel(grid; advection=WENO())
+ϵ(x, y, z) = rand()
+set!(model, u=ϵ, v=ϵ, w=ϵ)
+simulation = Simulation(model; Δt=0.01, stop_iteration=100)
```
Comment (simone-silvestri) [inline docs/src/field_time_series.md]: Using a little less iteration might help to reduce the burden on the documenter

File: docs/src/field_time_series.md [hunk 2]
```diff
@@ -1 +1,403 @@
...
+In place of an index, an `Oceananigans.Units.Time` can be used to interpolate values at a given time.
+
+```jldoctest field_time_series
+using Oceananingans.Units: Time
+
+fts[4, 4, 4, Time(0.33)]
+fts[Time(0.33)]
+```
+
+Contained data may be manipulated with [`set!`](@ref). Providing a single time index will simply pass [`set!`](@ref) to the corresponding `Field`:
```
Comment (simone-silvestri) [inline docs/src/field_time_series.md:121]: maybe we can also show that the same mechanism with `OnDisk` backends writes down the corresponding data to file?
(Maybe we need to introduce the concept of `backend` before though)

Comment (simone-silvestri) [reply docs/src/field_time_series.md:121]: Ah jut saw you wrote it later on! 🦾

File: docs/src/field_time_series.md [hunk 3]
```diff
@@ -1 +1,403 @@
...
+0.5 | 0.5   | 0.5    | 0.5
+1.0 | 1.0   | 1.0    | 1.0
+1.5 | 1.0   | 1.5    | 0.5
+2.0 | 1.0   | 2.0    | 0.0
+2.5 | 1.0   | 2.5    | 0.5
+3.0 | 1.0   | 3.0    | 1.0
+```
+
+## OnDisk FieldTimeSeries
+The default `backend` for a `FieldTimeSeries` is `InMemory`. This is sufficient for small timeseries, but large simulation data likely cannot all be loaded into memory. Instead, it is possible to lazily load a timeseries from storage by passing `backend = OnDisk()` to `FieldTimeSeries`:
```
Comment (simone-silvestri) [inline docs/src/field_time_series.md:199]: maybe we should also talk about `InMemory(N)` where only `N` time snapshots will be loaded in memory on a sliding window?

**Comment by simone-silvestri** [comment]
Looks very nice, thanks!

File: docs/src/field_time_series.md (hunk 3 above)
Comment (erin-atkinson) [reply docs/src/field_time_series.md:199]: I've added a note for `InMemory(N)`

File: docs/src/field_time_series.md (hunk 1 above)
Comment (erin-atkinson) [reply docs/src/field_time_series.md]: Okay. It was just copied and very unnecessary. I updated it to 10 iterations. Also, the initialisation is now just some vortex, no random fields.

**Comment by erin-atkinson** [comment]
If this is okay for a reviewer I'm happy for it to be merged.

**Review by simone-silvestri — APPROVED** [review]

File: docs/src/field_time_series.md [hunk 4]
```diff
@@ -1 +1,420 @@
...
+fts_linear = FieldTimeSeries{Center, Center, Center}(grid, [0, 1]; time_indexing=Linear())
+fts_cyclical = FieldTimeSeries{Center, Center, Center}(grid, [0, 1]; time_indexing=Cyclical())
+
+for fts in (fts_clamp, fts_linear, fts_cyclical)
+    set!(fts, 0, 1)
+    set!(fts, 1, 2)
+end
+```
+
+Each of the above `FieldTimeSeries` contain identical underlying data, however their behaviour differs when indexed using `Oceananigans.Units.Time` outside of the range of given times.
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
Each of the above `FieldTimeSeries` contain identical underlying data, however their behavior differs when indexed using `Oceananigans.Units.Time` outside of the range of given times.
```

File: docs/src/field_time_series.md [hunk 5]
```diff
@@ -1 +1,420 @@
...
+├── time_indexing: Linear()
+├── backend: InMemory()
+├── path: test.jld2
+├── name: u
+└── data: 14×14×14×11 OffsetArray(::Array{Float64, 4}, -2:11, -2:11, -2:11, 1:11) with eltype Float64 with indices -2:11×-2:11×-2:11×1:11
+    └── max=0.0, min=0.0, mean=0.0
+```
+
+## Extrapolated data
+`FieldTimeSeries` supports multiple methods of time indexing outside of the given times.
```
Comment (glwagner) [inline docs/src/field_time_series.md:156]:
```suggestion
## Extrapolated data

`FieldTimeSeries` supports multiple methods of time indexing outside of the given times.
```

File: docs/src/field_time_series.md [hunk 6]
```diff
@@ -1 +1,420 @@
...
+0.5 | 0.5   | 0.5    | 0.5
+1.0 | 1.0   | 1.0    | 1.0
+1.5 | 1.0   | 1.5    | 0.5
+2.0 | 1.0   | 2.0    | 0.0
+2.5 | 1.0   | 2.5    | 0.5
+3.0 | 1.0   | 3.0    | 1.0
+```
+
+## OnDisk FieldTimeSeries
+The default `backend` for a `FieldTimeSeries` is `InMemory`. This is sufficient for small timeseries, but large simulation data likely cannot all be loaded into memory. Instead, it is possible to lazily load a timeseries from storage by passing `backend = OnDisk()` to `FieldTimeSeries`:
```
Comment (glwagner) [inline docs/src/field_time_series.md:199]:
```suggestion
## OnDisk FieldTimeSeries

The default `backend` for a `FieldTimeSeries` is `InMemory`. This is sufficient for small timeseries, but large simulation data likely cannot all be loaded into memory. Instead, it is possible to lazily load a timeseries from storage by passing `backend = OnDisk()` to `FieldTimeSeries`:
```

File: docs/src/field_time_series.md [hunk 7]
```diff
@@ -1 +1,420 @@
...
+└── u: 8×8×8×11 FieldTimeSeries{InMemory} located at (Face, Center, Center) of u at test.jld2
+```
+
+An empty `FieldDataset` may also be constructed by providing a grid, saved times and a tuple of field names. Optional keyword arguments specify locations, indices, boundary conditions and output path (in the case of `backend = OnDisk()`):
+
+```jldoctest field_time_series
+grid = RectilinearGrid(; topology = (Periodic, Periodic, Bounded),
+                         size = (8, 8, 8),
+                         extent = (1, 1, 1)
+                      )
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
                         extent = (1, 1, 1))
```

File: docs/src/field_time_series.md [hunk 8]
```diff
@@ -1 +1,420 @@
...
+└── u: 8×8×8×11 FieldTimeSeries{InMemory} located at (Face, Center, Center) on CPU
+```
+
+A convenience constructor also exists to generate a `FieldDataset` according to a `NamedTuple` of preexisting fields. The following is equivalent to the above:
+
+```jldoctest field_time_series
+grid = RectilinearGrid(; topology = (Periodic, Periodic, Bounded),
+                         size = (8, 8, 8),
+                         extent = (1, 1, 1)
+                      )
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
                         extent = (1, 1, 1))
```

File: docs/src/field_time_series.md [hunk 9]
```diff
@@ -1 +1,420 @@
...
+└── data: 14×14×14 OffsetArray(view(::Array{Float64, 4}, :, :, :, 1), -2:11, -2:11, -2:11) with eltype Float64 with indices -2:11×-2:11×-2:11
+    └── max=10.9956, min=0.0, mean=5.49779
+```
+
+Providing a vector of `Field`s will copy the data in each `Field` to the `FieldTimeSeries`:
+
+```jldoctest field_time_series
+fields = [Field{Face, Center, Center}(grid) for n in 1:length(fts)]
+set!(fts, fields)
+fts
```
Comment (glwagner) [inline docs/src/field_time_series.md:140]:
```suggestion
set!(fts, fields)
```

`set!` should return the `fts`

File: docs/src/field_time_series.md [hunk 10]
```diff
@@ -1 +1,420 @@
...
+    Using `backend = InMemory(N)` will create a `FieldTimeSeries` that lazily loads windows of length `N` time steps. When indexed outside the currently loaded window, a window containing the requested time step and the following `N-1` time steps is loaded into memory.
+
+An empty `OnDisk` `FieldTimeSeries` can be created by also including `path` and `name` keywords. These can only be [`set!`](@ref) with a single `Field` and integer index. Doing so will write the data to storage if that index doesn't exist, creating a file if necessary.
+
+```jldoctest field_time_series
+new_ondisk_fts = FieldTimeSeries{Center, Center, Center}(grid, times;
+                                                         path = "new.jld2",
+                                                         name = "c",
+                                                         backend=OnDisk()
+                                                        )
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
                                                         backend=OnDisk())
```

File: docs/src/field_time_series.md [hunk 11]
```diff
@@ -1 +1,420 @@
...
+```jldoctest field_time_series
+grid = RectilinearGrid(; topology = (Periodic, Periodic, Bounded),
+                         size = (8, 8, 8),
+                         extent = (1, 1, 1)
+                      )
+times = 0:10
+fields = (:u, :v)
+location = (; u = (Face(), Center(), Center()),
+              v = (Center(), Face(), Center())
+            )
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
              v = (Center(), Face(), Center()))
```

File: docs/src/field_time_series.md [hunk 12]
```diff
@@ -1 +1,420 @@
...
+    # Read current timestep
+    set!(u, input_fds.u[i])
+    set!(v, input_fds.v[i])
+    set!(w, input_fds.w[i])
+
+    # Calculate fields
+    compute!(ke)
+
+    # Output
+    set!(output_fds, i; ke_density, ke)
```
Comment (glwagner) [inline docs/src/field_time_series.md]:
```suggestion
for n in 1:length(times)
    # Read current timestep
    set!(u, input_fds.u[n])
    set!(v, input_fds.v[n])
    set!(w, input_fds.w[n])

    # Calculate fields
    compute!(ke)

    # Output
    set!(output_fds, n; ke_density, ke)
```

**Review by glwagner — APPROVED** [review]
looks great! just some minor comments about formatting

**Comment by erin-atkinson** [comment]
I'm scared I'm missing something I need to do but if not this still needs to be merged.

**Comment by glwagner** [comment]
let's get the docs build green and then merge!

**Comment by erin-atkinson** [comment]
Okay, I think I got the doctests agnostic about the correct things, it runs fine on my computer. Let's see... This is a bit new to me.

**Comment by erin-atkinson** [comment]
Looks like there was a preexisting error in the docstring for setting field time series on disk also.

**Comment by glwagner** [comment]
thank you @erin-atkinson !!
